lib/gateway/service-map.js

```diff
@@ -0,0 +1,109 @@
+'use strict'
+const pmap = require('p-map')
+const URL = require('url').URL
+const {
+  isTypeExtensionNode,
+  isTypeDefinitionNode,
+  parse
+} = require('graphql')
+
+const buildRequest = require('./request')
+
+function createTypeMap (schemaDefinition) {
+  const parsedSchema = parse(schemaDefinition)
+  const typeMap = {}
+
+  for (const definition of parsedSchema.definitions) {
+    if (
+      isTypeDefinitionNode(definition) ||
+      (isTypeExtensionNode(definition) && definition.name.value === 'Query')
+    ) {
+      typeMap[definition.name.value] = new Set(definition.fields.map(field => field.name.value))
```

Comment:
I would avoid `map` and I'll iterate over `definition.fields` using a for loop that adds everything to the `Set` or you can use a `reduce` operation.